File: libs/east-node/packages/east-node-std/src/fetch.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import { East, StringType, IntegerType, BooleanType, DictType, OptionType, VariantType, NullType, StructType, BlobType, type ValueTypeOf, SortedMap } from "@elaraai/east";
import type { PlatformFunction } from "@elaraai/east/internal";
import { EastError } from "@elaraai/east/internal";

/**
 * HTTP method variant type.
 *
 * Represents the HTTP request method for fetch operations.
 * Supports standard HTTP methods: GET, POST, PUT, DELETE, PATCH, HEAD, OPTIONS.
 */
export const FetchMethod = VariantType({
    GET: NullType,
    POST: NullType,
    PUT: NullType,
    DELETE: NullType,
    PATCH: NullType,
    HEAD: NullType,
    OPTIONS: NullType,
});

/**
 * HTTP request configuration structure.
 *
 * Complete configuration for making HTTP requests with custom methods,
 * headers, and optional body content.
 */
export const FetchRequestConfig = StructType({
    url: StringType,
    method: FetchMethod,
    headers: DictType(StringType, StringType),
    body: OptionType(StringType),
});

/**
 * HTTP response structure.
 *
 * Complete HTTP response including status code, headers, body, and success indicator.
 * The `ok` field is true for status codes in the 200-299 range.
 */
export const FetchResponse = StructType({
    status: IntegerType,
    statusText: StringType,
    headers: DictType(StringType, StringType),
    body: StringType,
    ok: BooleanType,
});

/**
 * Performs an HTTP GET request and returns the response body as a string.
 *
 * Makes a simple GET request to the specified URL and returns the response body text.
 * This is a convenience function for basic GET requests without custom headers.
 * Throws an error if the response status is not in the 200-299 range.
 *
 * This is a platform function for the East language, enabling HTTP GET requests
 * in East programs running on Node.js.
 *
 * @param url - The URL to fetch (must be a valid HTTP/HTTPS URL)
 * @returns The response body as a string
 *
 * @throws {EastError} When request fails:
 * - Invalid URL
 * - Network error
 * - Non-2xx HTTP status code
 *
 * @example
 * ```ts
 * const fetchData = East.function([], StringType, $ => {
 *     const data = $.let(Fetch.get("https://api.example.com/data"));
 *     return data;
 * });
 * ```
 */
export const fetch_get = East.asyncPlatform("fetch_get", [StringType], StringType);

/**
 * Performs an HTTP GET request and returns the response body as binary data.
 *
 * Makes a simple GET request to the specified URL and returns the response body
 * as raw bytes (Blob/Uint8Array). Useful for downloading binary files like images,
 * PDFs, or database files.
 * Throws an error if the response status is not in the 200-299 range.
 *
 * This is a platform function for the East language, enabling binary HTTP GET requests
 * in East programs running on Node.js.
 *
 * @param url - The URL to fetch (must be a valid HTTP/HTTPS URL)
 * @returns The response body as binary data (Blob)
 *
 * @throws {EastError} When request fails:
 * - Invalid URL
 * - Network error
 * - Non-2xx HTTP status code
 *
 * @example
 * ```ts
 * const downloadFile = East.function([], BlobType, $ => {
 *     const data = $.let(Fetch.getBytes("https://example.com/file.bin"));
 *     return data;
 * });
 * ```
 */
export const fetch_get_bytes = East.asyncPlatform("fetch_get_bytes", [StringType], BlobType);

/**
 * Performs an HTTP POST request with a string body.
 *
 * Makes a POST request to the specified URL with the provided string body.
 * Sets Content-Type header to "text/plain" by default. For more control over
 * headers and methods, use {@link fetch_request}.
 * Throws an error if the response status is not in the 200-299 range.
 *
 * This is a platform function for the East language, enabling HTTP POST requests
 * in East programs running on Node.js.
 *
 * @param url - The URL to post to (must be a valid HTTP/HTTPS URL)
 * @param body - The request body as a string
 * @returns The response body as a string
 *
 * @throws {EastError} When request fails:
 * - Invalid URL
 * - Network error
 * - Non-2xx HTTP status code
 *
 * @example
 * ```ts
 * const postData = East.function([], StringType, $ => {
 *     const response = $.let(Fetch.post("https://api.example.com/submit", "data"));
 *     return response;
 * });
 * ```
 */
export const fetch_post = East.asyncPlatform("fetch_post", [StringType, StringType], StringType);

/**
 * Performs a full HTTP request with custom configuration.
 *
 * Makes an HTTP request with complete control over method, headers, and body.
 * Returns the full response including status, headers, and body. Unlike {@link fetch_get}
 * and {@link fetch_post}, this does not throw on non-2xx status codes - check the `ok`
 * field in the response instead.
 *
 * This is a platform function for the East language, enabling full HTTP requests
 * in East programs running on Node.js.
 *
 * @param config - The request configuration (url, method, headers, optional body)
 * @returns The complete HTTP response with status, headers, and body
 *
 * @throws {EastError} When request fails at the network level (not for HTTP errors)
 *
 * @example
 * ```ts
 * const makeRequest = East.function([], FetchResponse, $ => {
 *     const config = East.value({
 *         url: "https://api.example.com/data",
 *         method: variant("GET", null),
 *         headers: new Map([["Authorization", "Bearer token"]]),
 *         body: variant("none", null),
 *     }, FetchRequestConfig);
 *     return Fetch.request(config);
 * });
 * ```
 */
export const fetch_request = East.asyncPlatform("fetch_request", [FetchRequestConfig], FetchResponse);

/**
 * Node.js implementation of fetch platform functions.
 *
 * Pass this array to {@link East.compileAsync} to enable fetch operations.
 */
const FetchImpl: PlatformFunction[] = [
    fetch_get.implement(async (url: string) => {
        try {
            const response = await fetch(url);
            if (!response.ok) {
                throw new EastError(`HTTP ${response.status}: ${response.statusText}`, {
                    location: [{ filename: "fetch_get", line: 0n, column: 0n }]
                });
            }
            return await response.text();
        } catch (err: any) {
            if (err instanceof EastError) throw err;
            throw new EastError(`Failed to fetch ${url}: ${err.message}`, {
                location: [{ filename: "fetch_get", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),
    fetch_get_bytes.implement(async (url: string) => {
        try {
            const response = await fetch(url);
            if (!response.ok) {
                throw new EastError(`HTTP ${response.status}: ${response.statusText}`, {
                    location: [{ filename: "fetch_get_bytes", line: 0n, column: 0n }]
                });
            }
            const arrayBuffer = await response.arrayBuffer();
            return new Uint8Array(arrayBuffer);
        } catch (err: any) {
            if (err instanceof EastError) throw err;
            throw new EastError(`Failed to fetch ${url}: ${err.message}`, {
                location: [{ filename: "fetch_get_bytes", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),
    fetch_post.implement(async (url: string, body: string) => {
        try {
            const response = await fetch(url, {
                method: 'POST',
                body: body,
                headers: {
                    'Content-Type': 'text/plain',
                },
            });
            if (!response.ok) {
                throw new EastError(`HTTP ${response.status}: ${response.statusText}`, {
                    location: [{ filename: "fetch_post", line: 0n, column: 0n }]
                });
            }
            return await response.text();
        } catch (err: any) {
            if (err instanceof EastError) throw err;
            throw new EastError(`Failed to post to ${url}: ${err.message}`, {
                location: [{ filename: "fetch_post", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),
    fetch_request.implement(async (config: ValueTypeOf<typeof FetchRequestConfig>) => {
        const url = config.url;
        const method = config.method.type.toUpperCase(); // Get the variant tag
        const headers: Record<string, string> = {};

        // Convert Map to plain object
        for (const [key, value] of config.headers) {
            headers[key] = value;
        }

        try {
            const response = await fetch(url, {
                method: method as string,
                headers,
                body: config.body.value,
            });

            const responseHeaders = new SortedMap<string, string>();
            response.headers.forEach((value, key) => {
                responseHeaders.set(key, value);
            });

            const body = await response.text();

            return {
                status: BigInt(response.status),
                statusText: response.statusText,
                headers: responseHeaders,
                body,
                ok: response.ok,
            };
        } catch (err: any) {
            throw new EastError(`Failed to fetch ${url}: ${err.message}`, {
                location: [{ filename: "fetch_request", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),
];

/**
 * Grouped fetch platform functions.
 *
 * Provides HTTP client operations using the fetch API.
 *
 * @example
 * ```ts
 * import { East, StringType } from "@elaraai/east";
 * import { Fetch } from "@elaraai/east-node-std";
 *
 * const getData = East.function([], StringType, $ => {
 *     const response = $.let(Fetch.get("https://api.example.com/data"));
 *     return response;
 * });
 *
 * const compiled = await East.compileAsync(getData.toIR(), Fetch.Implementation);
 * await compiled();  // Returns response body as string
 * ```
 */
export const Fetch = {
    /**
     * Performs an HTTP GET request and returns the response body.
     *
     * Makes a simple GET request to the specified URL and returns the response body text.
     * Throws an error if the response status is not in the 200-299 range.
     *
     * @param url - The URL to fetch
     * @returns The response body as a string
     * @throws {EastError} When request fails or status is not 2xx
     *
     * @example
     * ```ts
     * const fetchData = East.function([], StringType, $ => {
     *     return Fetch.get("https://api.example.com/data");
     * });
     *
     * const compiled = await East.compileAsync(fetchData.toIR(), Fetch.Implementation);
     * await compiled();  // Returns: response body as string
     * ```
     */
    get: fetch_get,

    /**
     * Performs an HTTP GET request and returns the response body as binary data.
     *
     * Makes a simple GET request to the specified URL and returns the response body
     * as raw bytes (Blob/Uint8Array). Useful for downloading binary files.
     * Throws an error if the response status is not in the 200-299 range.
     *
     * @param url - The URL to fetch
     * @returns The response body as binary data (Blob)
     * @throws {EastError} When request fails or status is not 2xx
     *
     * @example
     * ```ts
     * const downloadFile = East.function([], BlobType, $ => {
     *     return Fetch.getBytes("https://example.com/file.bin");
     * });
     *
     * const compiled = await East.compileAsync(downloadFile.toIR(), Fetch.Implementation);
     * await compiled();  // Returns: binary data as Uint8Array
     * ```
     */
    getBytes: fetch_get_bytes,

    /**
     * Performs an HTTP POST request with a string body.
     *
     * Makes a POST request to the specified URL with the provided string body.
     * Sets Content-Type to "text/plain" by default.
     * Throws an error if the response status is not in the 200-299 range.
     *
     * @param url - The URL to post to
     * @param body - The request body as a string
     * @returns The response body as a string
     * @throws {EastError} When request fails or status is not 2xx
     *
     * @example
     * ```ts
     * const postData = East.function([], StringType, $ => {
     *     return Fetch.post("https://api.example.com/submit", "data");
     * });
     *
     * const compiled = await East.compileAsync(postData.toIR(), Fetch.Implementation);
     * await compiled();  // Returns: response body as string
     * ```
     */
    post: fetch_post,

    /**
     * Performs a full HTTP request with custom configuration.
     *
     * Makes an HTTP request with complete control over method, headers, and body.
     * Returns the full response including status, headers, and body.
     * Does not throw on non-2xx status - check the `ok` field instead.
     *
     * @param config - The request configuration
     * @returns The complete HTTP response
     * @throws {EastError} When request fails at the network level
     *
     * @example
     * ```ts
     * const makeRequest = East.function([], FetchResponse, $ => {
     *     const config = East.value({
     *         url: "https://api.example.com/data",
     *         method: variant("GET", null),
     *         headers: new Map(),
     *         body: variant("none", null),
     *     }, FetchRequestConfig);
     *     return Fetch.request(config);
     * });
     *
     * const compiled = await East.compileAsync(makeRequest.toIR(), Fetch.Implementation);
     * const response = await compiled();  // Returns: FetchResponse object
     * ```
     */
    request: fetch_request,

    /**
     * Node.js implementation of fetch platform functions.
     *
     * Pass this to {@link East.compileAsync} to enable fetch operations.
     */
    Implementation: FetchImpl,

    /**
     * Type definitions for fetch operations.
     */
    Types: {
        /**
         * HTTP method variant type (GET, POST, PUT, DELETE, PATCH, HEAD, OPTIONS).
         */
        Method: FetchMethod,

        /**
         * HTTP request configuration structure.
         */
        RequestConfig: FetchRequestConfig,

        /**
         * HTTP response structure.
         */
        Response: FetchResponse,
    },
} as const;

// Export for backwards compatibility
export { FetchImpl };

File: libs/east-py/packages/east-py-std/east_py_std/fetch.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""HTTP fetch platform functions for East.

Provides HTTP request operations for East programs running in Python.
The ``*_impl`` functions are plain Python callables taking and returning East
values - import them directly from a project's own ``@platform_function`` to
reuse the implementations without an IR round-trip. The East type definitions
are re-exported here for building inputs with ``coerce_to`` and validating
outputs.
"""

import urllib.error
import urllib.request
from typing import TypedDict, cast

from east.runtime.platform import platform_function, platform_functions
from east.types.types import (
    BlobType,
    BooleanType,
    DictType,
    IntegerType,
    NullType,
    OptionType,
    StringType,
    StructType,
    VariantType,
)
from east.types.values import EastBlob, EastDict, EastStruct, EastVariant


class HttpMethodVariant(TypedDict):
    """HTTP method variant structure."""

    type: str  # "GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"
    value: None


class OptionStringVariant(TypedDict):
    """Option variant structure for optional string."""

    type: str  # "some" or "none"
    value: str | None


class FetchRequestConfig(TypedDict):
    """HTTP request configuration structure."""

    url: str
    method: EastVariant[HttpMethodVariant]
    headers: EastDict
    body: EastVariant[OptionStringVariant]


class FetchResponse(TypedDict):
    """HTTP response structure."""

    status: int
    statusText: str
    headers: EastDict
    body: str
    ok: bool


# ============================================================================
# Type Definitions for Fetch
# ============================================================================

FetchMethodType = VariantType(
    [
        ("GET", NullType),
        ("POST", NullType),
        ("PUT", NullType),
        ("DELETE", NullType),
        ("PATCH", NullType),
        ("HEAD", NullType),
        ("OPTIONS", NullType),
    ]
)
"""HTTP method selector.

Cases: ``GET``, ``POST``, ``PUT``, ``DELETE``, ``PATCH``, ``HEAD``,
``OPTIONS``.
"""

FetchRequestConfigType = StructType(
    [
        ("url", StringType),
        ("method", FetchMethodType),
        ("headers", DictType(StringType, StringType)),
        ("body", OptionType(StringType)),
    ]
)
"""HTTP request configuration for ``fetch_request``.

Fields: ``url`` (``String``), ``method`` (``FetchMethodType``),
``headers`` (``Dict<String, String>``), ``body``
(``Option<String>`` - request body; ``none`` for bodyless methods).
"""

FetchResponseType = StructType(
    [
        ("status", IntegerType),
        ("statusText", StringType),
        ("headers", DictType(StringType, StringType)),
        ("body", StringType),
        ("ok", BooleanType),
    ]
)
"""HTTP response returned by ``fetch_request``.

Fields: ``status`` (``Integer`` - HTTP status code), ``statusText``
(``String`` - reason phrase), ``headers`` (``Dict<String, String>`` -
lowercase-normalised response headers), ``body`` (``String`` - UTF-8
decoded response body), ``ok`` (``Boolean`` - ``True`` when
200 <= status < 300).
"""


@platform_function(name="fetch_get", inputs=[StringType], output=StringType)
async def fetch_get_impl(url: str) -> str:
    """Perform an HTTP GET request and return the response body.

    Args:
        url: ``String`` (``str``) - URL to fetch.

    Returns:
        ``String`` (``str``) - UTF-8 decoded response body.

    Raises:
        Exception: If the request fails or the HTTP status is not 2xx.
    """
    import asyncio

    loop = asyncio.get_event_loop()

    def _fetch() -> str:
        response = urllib.request.urlopen(url)
        if response.status < 200 or response.status >= 300:
            raise Exception(f"HTTP {response.status}: {response.msg}")
        return response.read().decode("utf-8")

    return await loop.run_in_executor(None, _fetch)


@platform_function(name="fetch_get_bytes", inputs=[StringType], output=BlobType)
async def fetch_get_bytes_impl(url: str) -> EastBlob:
    """Perform an HTTP GET request and return the response body as raw bytes.

    Args:
        url: ``String`` (``str``) - URL to fetch.

    Returns:
        ``Blob`` (``EastBlob``) - raw response body bytes.

    Raises:
        Exception: If the request fails or the HTTP status is not 2xx.
    """
    import asyncio

    loop = asyncio.get_event_loop()

    def _fetch() -> EastBlob:
        response = urllib.request.urlopen(url)
        if response.status < 200 or response.status >= 300:
            raise Exception(f"HTTP {response.status}: {response.msg}")
        return EastBlob(response.read())

    return await loop.run_in_executor(None, _fetch)


@platform_function(name="fetch_post", inputs=[StringType, StringType], output=StringType)
async def fetch_post_impl(url: str, body: str) -> str:
    """Perform an HTTP POST request with a plain-text body.

    Sends ``body`` encoded as UTF-8 with ``Content-Type: text/plain``.

    Args:
        url: ``String`` (``str``) - URL to POST to.
        body: ``String`` (``str``) - request body.

    Returns:
        ``String`` (``str``) - UTF-8 decoded response body.

    Raises:
        Exception: If the request fails or the HTTP status is not 2xx.
    """
    import asyncio

    loop = asyncio.get_event_loop()

    def _fetch() -> str:
        req = urllib.request.Request(
            url, data=body.encode("utf-8"), headers={"Content-Type": "text/plain"}, method="POST"
        )
        response = urllib.request.urlopen(req)
        if response.status < 200 or response.status >= 300:
            raise Exception(f"HTTP {response.status}: {response.msg}")
        return response.read().decode("utf-8")

    return await loop.run_in_executor(None, _fetch)


@platform_function(name="fetch_request", inputs=[FetchRequestConfigType], output=FetchResponseType)
async def fetch_request_impl(config: EastStruct[FetchRequestConfig]) -> EastStruct[FetchResponse]:
    """Perform an HTTP request with full control over method, headers, and body.

    Non-2xx responses are captured rather than raised: check ``ok`` or ``status``
    on the returned struct. Network-level errors (DNS failure, connection refused,
    timeout) are raised as exceptions.

    Args:
        config: ``FetchRequestConfigType`` (``EastStruct``) with fields:

            - ``url`` (``String``): request URL.
            - ``method`` (``FetchMethodType``): HTTP verb (``GET``, ``POST``,
              ``PUT``, ``DELETE``, ``PATCH``, ``HEAD``, ``OPTIONS``).
            - ``headers`` (``Dict<String, String>``): additional request headers.
            - ``body`` (``Option<String>``): request body encoded as UTF-8;
              use ``none`` for bodyless methods.

    Returns:
        ``FetchResponseType`` (``EastStruct``): ``status`` (``Integer``),
        ``statusText`` (``String``), ``headers`` (``Dict<String, String>``
        with lowercase keys), ``body`` (``String``), ``ok`` (``Boolean``).

    Raises:
        Exception: On network-level failure (DNS, connection, timeout).
    """
    import asyncio

    loop = asyncio.get_event_loop()

    def _fetch() -> EastStruct[FetchResponse]:
        # Cast to TypedDict for type inference (workaround until EastStruct.__getitem__ is typed)
        cfg = cast(FetchRequestConfig, config)

        # Extract config fields
        url = cfg["url"]
        method_variant = cast(HttpMethodVariant, cfg["method"])
        headers_dict = cfg["headers"]
        body_option = cast(OptionStringVariant, cfg["body"])

        # Convert method variant to string
        method = method_variant["type"]

        # Convert headers dict to Python dict
        headers = dict(headers_dict.items())

        # Get body if present
        data = None
        if body_option["type"] == "some":
            body_str = body_option["value"]
            if body_str is not None:
                data = body_str.encode("utf-8")

        # Create request
        req = urllib.request.Request(url, data=data, headers=headers, method=method)

        # Execute request
        try:
            response = urllib.request.urlopen(req)
            status = response.status
            status_text = response.msg
            body = response.read().decode("utf-8")
            response_headers = dict(response.headers)
        except urllib.error.HTTPError as e:
            status = e.code
            status_text = e.msg
            body = e.read().decode("utf-8") if e.fp else ""
            response_headers = dict(e.headers)

        # Convert response headers to EastDict (lowercase keys for consistency)
        east_headers: EastDict = EastDict(
            StringType, StringType, {k.lower(): v for k, v in response_headers.items()}
        )

        # Create response struct (plain dict wrapped in EastStruct)
        return EastStruct(
            {
                "status": status,
                "statusText": status_text,
                "headers": east_headers,
                "body": body,
                "ok": (200 <= status < 300),
            }
        )

    return await loop.run_in_executor(None, _fetch)


# Collected from the @platform_function decorations above.
fetch_impl = platform_functions(__name__)


__all__ = [
    "fetch_impl",
    "fetch_get_impl",
    "fetch_get_bytes_impl",
    "fetch_post_impl",
    "fetch_request_impl",
    "FetchMethodType",
    "FetchRequestConfigType",
    "FetchResponseType",
]
